### PR TITLE
backup: use history iterator to scan ddl jobs

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/54139

Problem Summary:
sometimes cluster might have too many ddl jobs. backup should skip unnecessary scan of ddl jobs during the backup incremental ddls.
### What changed and how does it work?
Use `GetLastHistoryDDLJobsIterator` instead of `GetAllDDLJobs`.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -1000,6 +1000,38 @@ func WriteBackupDDLJobs(metaWriter *metautil.MetaWriter, g glue.Glue, store kv.S
...
+				if err != nil {
+					return true, errors.Trace(err)
+				}
+				err = metaWriter.Send(jobBytes, metautil.AppendDDL)
+				if err != nil {
+					return true, errors.Trace(err)
+				}
+			}
+			// early exits to stop unnecessary scan
+			return true, nil
```
Comment (kennytm) [inline br/pkg/backup/client.go:1030]: The early exit condition is wrong. The `if` condition on line 1010 can fail in cases like `state != Done` or `SchemVersion > backupSchemaVersion` which you need to continue scanning.

The only good stop condition is `SchemaVersion <= lastSchemaVersion`.

File: br/pkg/backup/client.go [hunk 2]
```diff
@@ -1014,41 +1044,49 @@ func WriteBackupDDLJobs(metaWriter *metautil.MetaWriter, g glue.Glue, store kv.S
...
+	for {
+		cacheJobs, err = historyJobsIter.GetLastJobs(ddl.DefNumHistoryJobs, cacheJobs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(cacheJobs) == 0 {
+			// no more jobs
+			break
+		}
+		count += len(cacheJobs)
```
Comment (Leavrth) [inline br/pkg/backup/client.go:1067]: should `count` counts the filtered jobs? `count += len(jobs)`

**Review by Leavrth — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**